File: python/98_validate_binary_search_tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.isValidHelper(root, None, None)
    
    def isValidHelper(self, root, minVal, maxVal):
        """
        :type root: TreeNode
        :type minVal, maxVal: int
        :rtype: bool
        """
        if root:
            if minVal and root.val <= minVal.val:
                return False
            if maxVal and root.val >= maxVal.val:
                return False
            return self.isValidHelper(root.left, minVal, root) and \
                   self.isValidHelper(root.right, root, maxVal)
        else:
            return True

File: c++/98_validate_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool isValidBSTHelper(TreeNode* root, long minVal, long maxVal) {
        if (root != nullptr) {
            if (root->val <= minVal) {
                return false;
            } else if (root->val >= maxVal){
                return false;
            } else {
                return (isValidBSTHelper(root->left, minVal, root->val) &&
                        isValidBSTHelper(root->right, root->val, maxVal));
            }
        } else {
            return true;
        }
    }

public:
    bool isValidBST(TreeNode* root) {
        long min = static_cast<long> (INT_MIN);
        long max = static_cast<long> (INT_MAX);
        return isValidBSTHelper(root, min - 1, max + 1);
    }
};
